Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp

```diff
@@ -0,0 +1,1053 @@
+//   in HTMLElement::beginParsingChildren(), which this does not call.
+template<class Char>
+class HTMLFastPathParser {
+    using CharSpan = Span<const Char>;
+    using UCharSpan = Span<const UChar>;
+    static_assert(std::is_same_v<Char, UChar> || std::is_same_v<Char, LChar>);
+
+public:
+    HTMLFastPathParser(CharSpan source, Document& document, DocumentFragment& fragment)
+        : m_source(source)
+        , m_document(document)
+        , m_fragment(fragment)
+    {
+    }
+
+    bool parse(Element& contextElement)
+    {
+        auto contextTag = contextElement.tagQName();
+        ASSERT(!contextTag.localName().isEmpty());
+
+        // This switch checks that the context element is supported and applies the
+        // same restrictions regarding content as the fast-path parser does for a
+        // corresponding nested tag.
+        // This is to ensure that we preserve correct HTML structure with respect
+        // to the context tag.
+        //
+        // If this switch has duplicate cases, then `tagNameHash()` needs to be
+        // updated.
+        switch (tagNameHash(contextTag.localName())) {
+#define TAG_CASE(TagName, TagClassName)                                                                      \
+        case tagNameHash(TagInfo::TagClassName::tagName):                                                    \
+            ASSERT(HTMLNames::TagName##Tag->localName().string().ascii() == TagInfo::TagClassName::tagName); \
+            if constexpr (!TagInfo::TagClassName::isVoid) {                                                  \
+                /* The hash function won't return collisions for the supported tags, but this function */    \
+                /* takes potentially unsupported tags, which may collide. Protect against that by */         \
+                /* checking equality. */                                                                     \
+                if (contextTag == HTMLNames::TagName##Tag) {                                                 \
+                    parseCompleteInput<typename TagInfo::TagClassName>();                                    \
+                    return !m_parsingFailed;                                                                 \
+                }                                                                                            \
+            }                                                                                                \
+        break;
+        FOR_EACH_SUPPORTED_TAG(TAG_CASE)
+        default:
+            break;
+#undef TAG_CASE
+        }
+
+        didFail(HTMLFastPathResult::FailedUnsupportedContextTag);
+        return false;
+    }
+
+    HTMLFastPathResult parseResult() const { return m_parseResult; }
+
+private:
+    CharSpan m_source;
+    Document& m_document;
+    DocumentFragment& m_fragment;
+
+    const Char* const m_end { m_source.data() + m_source.size() };
```

Comment:
Did you consider using an abstraction like the StringParsingBuffer rather than raw pointers?

Reply:
This is a straight import. I didn't make any tweaks yet. I think we should adopt for WebKit concepts, though I was hoping to do it in a follow-up since each change will have to be perf-tested.

Reply:
I'll give it a shot in a follow-up and perf test it. I would be nicer than dealing with pointers indeed.

Reply:
I was able to rewrite the code using StringParsingBuffer easily. I am waiting on A/B perf results.

Reply:
Looks like a small regression at the moment, I am trying to figure out why.